### PR TITLE
Buttons: Add variations for vertical layout

### LINKED ISSUE #20081: Add "Vertical" block style to buttons block
**Is your feature request related to a problem? Please describe.**
Currently the default styles of buttons blocks results in its 
containing button blocks being aligned horizontally.
This can be annoying if existing, singular button blocks (from a time where Gutenberg hadn't used a wrapper buttons block when a button block is inserted) are [transformed to be wrapped by a buttons block](https://github.com/WordPress/gutenberg/issues/19919#issuecomment-583182682), because singular button blocks on each line would suddenly put on a single line in the containing buttons block.

It would therefore make sense to either add a "Vertical" core block style for the buttons block or style the buttons block default styles to vertically aligned button blocks right away.

**Describe the solution you'd like**
"Vertical" block style for buttons block -or-
Default blocks styles align its containing button blocks vertically, one on each line.

**Describe alternatives you've considered**
The alternative would be to add such a "vertical" buttons block style by plugin/theme or re-styling the default buttons block style by plugin/theme, or adding some inline styles. This can be a hassle for users that migrate singular button blocks to wrapped button blocks in buttons blocks.

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Resolves: https://github.com/WordPress/gutenberg/issues/20081

This PR adds block variations to `Buttons` block for enabling the switch to `vertical` and `horizontal` orientation, in the same fashion `Navigation` block handles this.

## How has this been tested?
Manually and by ensuring that previous versions of Buttons blocks would remain unaffected by this change.

## Screenshots 
verticalButtons

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by guarani** [comment]
Smoke tested on react-native-editor and there are no regressions there. If in the future we want to add this to mobile, it can be added.

**Comment by talldan** [comment]
This is mostly working really nicely, though the spacing between blocks looks different in my local env to your screen capture:
Screenshot 2020-11-27 at 12 55 31 pm

Not sure if this is a theme issue, it does look better in Twenty Nineteen and Twenty Twenty One Blocks, but the spacing is bigger in Twenty Twenty and Twenty Twenty One (non-blocks).

The 'Use Theme Styles' toggle doesn't seem to be working anymore, so that's making it hard to check.

**Review by gziolo — APPROVED** [review]
Technical implementation of the block looks good. It works similar to how Navigation block allows transforming its variations. `save` method was updated but it should produce the same output for existing content.

Let's make sure that the comment about CSS from @talldan is addressed before proceeding. @mapk or @jasmussen can you help with that?

**Comment by gziolo** [comment]
It looks like integration tests for the Buttons block need to be regenerated to work with the new `orientation` attribute added.

**Comment by ntsekouras** [comment]
>This is mostly working really nicely, though the spacing between blocks looks different in my local env to your screen capture

Hey @talldan - thanks for reviewing!

This is something that can be affected by theme styles but your screenshot has big gaps :) . I tested with 2021(both blocks and non-blocks) and 2019 and seem quite good. In 2020 there are theme styles that cause that (margin top and bottom are 30px). Also previously the `Buttons` block had bigger `top-margins` that I removed it in this PR to have the same as every block. Maybe `css` was cached in browser? Do you mind checking once again and make sure the new css is loaded?

>The 'Use Theme Styles' toggle doesn't seem to be working anymore, so that's making it hard to check.

What is this toggle?

**Comment by talldan** [comment]
> What is this toggle?

In the editor preferences
Screenshot 2020-11-27 at 4 34 20 pm

Pretty useful for testing whether the default styling looks ok!

I'll give it a re-test to see what's happening with the margins.

**Comment by talldan** [comment]
@ntsekouras It's this style in Twenty Twenty One non-blocks that's causing the gap in my local environment:
https://github.com/WordPress/twentytwentyone/blob/4c83b2781ae9cc9002ef8e0a5c275b2cea8a7f08/assets/sass/05-blocks/utilities/_editor.scss#L73-L76

I wonder if we're using different versions of the theme. Can you see this style if you inspect the button block wrapper?
Screenshot 2020-11-27 at 4 50 39 pm

It might not be something to worry about in the core block styles, but perhaps we should make an issue over on the theme repo if we can validate the issue:
https://github.com/WordPress/twentytwentyone

**Comment by ntsekouras** [comment]
This is what I see..

Screenshot 2020-11-27 at 11 03 48 AM

>It might not be something to worry about in the core block styles, but perhaps we should make an issue over on the theme repo if we can validate the issue 

Since is not merged I could ping some people who work on that theme to validate. I'll do that now.

**Comment by jasmussen** [comment]
This is what I see:

movers

My immediate instinct was that vertical or horizontal should just be style variations for the block. However as Nik kindly explained — the challenge with this block is the same as it is for the navigation block: *style variations are just CSS changes, but these blocks need their horizontal movers swapped out with vertical movers, which require a different kind of transform*.

Because that is the case, this approach is okay. 

---

However, and not necessarily as part of this PR, but we really need to improve this selector:

Screenshot 2020-11-27 at 10 29 53

It looks like a dropdown, but it doesn't act like a dropdown. It doesn't know its own transform, and it doesn't show it inside the dropdown field, as it should. Also because of that, when you click it the first time, there's no checkbox, which suggests you can transform a horizontal block to a horizontal block. It's not a good UI, and we really should improve it before we expand it that much further.

Secondly, I wish there was a way to revisit the mover control so it could know itself whether to be vertical or horizontal, and avoid having us build these UIs on top of it. Could it simply read the flex-direction of its parent?

**Comment by talldan** [comment]
@ntsekouras The screenshot looks like the styles for the buttons block. How about the styles for an individual button?

**Comment by ntsekouras** [comment]
>The screenshot looks like the styles for the buttons block. How about the styles for an individual button?

Screenshot 2020-11-27 at 11 55 01 AM

I saw now that the margin applies on selecting the `Buttons`. I do believe this is for themes to style though, since is a new layout? This affects all themes and I think that every new thing like that would cause themes to handle the styles, no?

**Comment by scruffian** [comment]
This is great, thank you!

File: packages/block-library/src/button/editor.scss [hunk 1]
```diff
@@ -19,7 +19,7 @@
 	}
 
 	// Add outline to button on focus to indicate focus-state
-	&:focus {
+	&:focus:not(.wp-block-buttons &) {
```
Comment (ZebulanStanphill) [inline packages/block-library/src/button/editor.scss:22]: What's the purpose of this change?

File: packages/block-library/src/buttons/style.scss [hunk 2]
```diff
@@ -18,10 +29,16 @@
 
 	&.is-content-justification-left {
```
Comment (ZebulanStanphill) [inline packages/block-library/src/buttons/style.scss:30]: Perhaps this className should have been `is-content-justification-start`, because "left" doesn't make sense in the vertical context, and the actual value of `align-items` that gets set is `flex-start` anyway. Same for `is-content-justification-right` which should probably be `is-content-justification-end`.

**Review by ZebulanStanphill — DISMISSED** [review]
~Let's add an `orientation` prop to the `<ContentJustificationDropdown>` component, in order to provide a way to change the options shown to match the orientation. (Technically it would just rename them, I guess.)~

~In hindsight, when I implemented the content justification controls, I think I should have gone with `start` and `end` terminology for the variable/prop names rather than `left` and `right`, which don't make sense in a vertical context. Let's try and fix that while we're at it.~

**Comment by talldan** [comment]
`ContentJustificationDropdown` only handles horizontal alignments, so I don't really see an issue. The buttons block can't be made bigger than the content, so I don't think there's a need to align vertically right now.

File: packages/block-library/src/buttons/style.scss (hunk 2 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: Just realized the above comment was incorrect, since I got horizontal/vertical controls confused in my head.

Anyway, there is an actual problem here, though: this class is being used to `align-items` for vertical Buttons blocks, but the class name implies it can only set `justify-content`. Additionally, the `justify-content` styles are being applied to both vertical and horizontal orientations. That's not technically correct.

**Comment by ZebulanStanphill** [comment]
@talldan Oh, right. Sorry, I got mixed up about what was going on in this PR.

Taking a closer look, it's kind of weird that the component called `ContentJustificationDropdown` is actually setting `align-items` when the block is using the vertical orientation. And the same goes for the classes. I wonder what the proper term would be to refer to a setting that can change either `align-items` or `justify-content` depending on the context.

File: packages/block-library/src/button/editor.scss (hunk 1 above)
Comment (ntsekouras) [reply packages/block-library/src/button/editor.scss:22]: If you check on master, when you focus on a button it shows both outline and box-shadow and it's really weird. It's not related with these changes but makes it look like the other blocks.

File: packages/block-library/src/buttons/style.scss (hunk 2 above)
Comment (ntsekouras) [reply packages/block-library/src/buttons/style.scss:30]: Why not? You can have vertically stacked buttons and align them for example to the right. What do you propose exactly?

Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: We shouldn't be setting `justify-content` for the vertical case... only `align-items`. But also, the class "is-content-justification-xxx" is kind of confusing if it's actually applying item alignment in the vertical cases. I think ideally, we'd want to tackle this stuff sooner rather than later to reduce the number of deprecations and legacy styles we have to keep on the front-end.

File: packages/block-library/src/button/editor.scss (hunk 1 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/button/editor.scss:22]: Okay, but why is this change only being applied to Button blocks nested inside a Buttons block? What makes Button blocks outside of a Buttons block different from those nested in one? And perhaps a comment here would help explain this change, since it's not clear from just reading the CSS.

File: packages/block-library/src/buttons/style.scss (hunk 2 above)
Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: The problem here is basically that I named the `is-content-justification-xxx` classes with the assumption that they would only be used to apply different `justify-content` styles, but that's no longer true in this case.

Maybe a new class name that can work both cases would be appropriate. I'm not sure what the correct term is, though.

Or maybe the solution is to apply different classes for the vertical cases, e.g. `is-item-alignment-xxx`. I'm not sure.

File: packages/block-library/src/button/editor.scss (hunk 1 above)
Comment (jasmussen) [reply packages/block-library/src/button/editor.scss:22]: Because the Button (singular) block has an input field, it's annoyingly hard to set focus on the Button directly, but it is possible. Here's how that looks in the main branch:

master

Here's how that looks in this branch:

this branch

The change seems good. I'd agree it'd be nice to be able to avoid the :not selector. Could a rule exist in the Buttons block editor style instead?

File: packages/block-library/src/buttons/style.scss (hunk 2 above)
Comment (jasmussen) [reply packages/block-library/src/buttons/style.scss:30]: This is gnarly, and it seems like there are several things going on:

- The desire for a more abstracted system of defining flex containers (CC: @youknowriad as I know you've thought about this)
- The fact that in flex, the two properties `align-items` and `justify-content` center vertically and horizontally, depending on which `flex-direction` is applied.

While both of those are valid, I think maybe we're establishing properties here that don't have anything to do with the underlying tech, which could be flex, or CSS grid, or something else. In that vein, personally I think of _alignment_ as the positioning of the block itself, in this case the container block, and _justificiation_ as the positioning of content inside. 

Is that the path forward? Deciding on some classnames that have nothing to do with the CSS that is used, but only with the editor visual semantics?

Comment (youknowriad) [reply packages/block-library/src/buttons/style.scss:30]: Agreed here, I don't think we should be naming classnames based on technology but more based on intent and yes it seems we want to "align content" with this. It's called now "justify content" which happens to map with "justify-content" on the default flex orientation. I don't really care too much that it doesn't match the naming of the  CSS property in vertical orientation and I don't really think we should be renaming the classname just for that.

Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: > personally I think of _alignment_ as the positioning of the block itself, in this case the container block, and _justificiation_ as the positioning of content inside.

It's worth remembering that in addition to `align-items` and `justify-content`, there's also [`align-content`](https://developer.mozilla.org/en-US/docs/Web/CSS/align-content) and [`justify-items`](https://developer.mozilla.org/en-US/docs/Web/CSS/justify-items), though `justify-items` doesn't do anything in Flexbox.

By looking at how these properties work, we can learn the following about how CSS is using these terms:

In a horizontally-flowing container, "align" refers to vertical positioning and "justify" refers to horizontal positioning. In a vertically-flowing container, "align" refers to horizontal positioning and "justify" refers to vertical positioning. In other words, "justify" refers to positioning along the primary axis that content is flowing, and "align" refers to the secondary (AKA "cross") axis.

"content" applies to all the children as a single block within the container, and "items" applies to each child as an individual unit.

Technically, the kind of alignment/justification we want for the horizontal variant is actually conceptually different from the control we want on the vertical variant. Technically, `justify-content`'s cross-axis complement is `align-content`, not `align-items`. :confounded:

Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: If I had to come up with a better class name than what we're using right now, I guess something like `is-content-horizontal-placement-xxx` might work. I'm drawing the "placement" term here from the shorthand properties [`place-content`](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content) and [`place-items`](https://developer.mozilla.org/en-US/docs/Web/CSS/place-items). The CSS "placement" term is apparently its generic way of referring to both "alignment" and "justification".

Of course, in my suggested class name here, the meaning of "content" is slightly different from what CSS uses it for, since in the case of my suggested class name, it can refer to either CSS's "content" or "items". But I guess it's "good enough" for our purposes?

Comment (youknowriad) [reply packages/block-library/src/buttons/style.scss:30]: For me it's too late to change the class names, this is already shipped and themes might be relying on it. The renaming doesn't seem to add a lot of value.

Comment (ZebulanStanphill) [reply packages/block-library/src/buttons/style.scss:30]: As far as I can tell, it's only shipped in the plugin, not core, right? Does that change anything?

Comment (youknowriad) [reply packages/block-library/src/buttons/style.scss:30]: > Does that change anything?

It does, we have more luxury to change things that way but are we certain of that? It's also a bit late if it's in WP 5.6 as we're on RC2 and the stable is supposed to happen next week. 

We need to do a few things if we want to rename the classnames:

 - try to check usage of the classname in the plugins/theme repository
 - backport the fix to  WP 5.6 (so it needs to be in its own separate PR and not block this one)

cc @tellthemachines

**Comment by ntsekouras** [comment]
Based on the above discussion about `css classes` (https://github.com/WordPress/gutenberg/pull/27297#discussion_r532313021), I'm going to merge this.

The next steps will be determined by:
>We need to do a few things if we want to rename the classnames:
> - try to check usage of the classname in the plugins/theme repository
> - backport the fix to WP 5.6 (so it needs to be in its own separate PR and not block this one)
